runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -5118,172 +5118,277 @@ TR_PrexArgInfo* TR_PrexArgInfo::argInfoFromCaller(TR::Node* callNode, TR_PrexArg
+      }
 
-      TR_PrexArgInfo* argsFromCaller = TR_PrexArgInfo::argInfoFromCaller(site->_callNode, callerArgInfo);
-      target->_prexArgInfo = TR_PrexArgInfo::enhance(target->_prexArgInfo, argsFromCaller, comp());
+   // At this stage, we can improve the type of the virtual guard we are going to use
+   // For a non-overridden guard or for an interface guard, if it makes sense, try to use a vft-test
+   // based virtual guard
+   //
+   bool disablePreexForChangedGuard = false;
+   TR_PersistentCHTable * chTable = comp()->getPersistentInfo()->getPersistentCHTable();
+   TR_PersistentClassInfo *thisClassInfo = chTable->findClassInfoAfterLocking(target->_receiverClass, comp());
+   if (target->_calleeSymbol->hasThisCalls() &&
+       target->_receiverClass &&
+       !TR::Compiler->cls.isAbstractClass(comp(), target->_receiverClass) &&
+       !fe()->classHasBeenExtended(target->_receiverClass) &&
+       thisClassInfo &&
+       thisClassInfo->isInitialized() &&
+       ((target->_guard->_kind == TR_NonoverriddenGuard && target->_guard->_type == TR_NonoverriddenTest) ||
+        (target->_guard->_kind == TR_InterfaceGuard)) &&
+       performTransformation(comp(), "O^O VIRTUAL GUARD IMPROVE: Changed guard kind %s type %s to use VFT test\n", tracer()->getGuardKindString(target->_guard), tracer()->getGuardTypeString(target->_guard)))
+      {
+      target->_guard->_type = TR_VftTest;
+      target->_guard->_thisClass = target->_receiverClass;
       }
 
-   return computePrexInfo(target);
+   return prexArgInfo;
    }
 
 TR_PrexArgInfo *
 TR_J9InlinerUtil::computePrexInfo(TR_CallTarget *target)
    {
-   TR_CallSite *site = target->_myCallSite;
-   if (!site)
+   return computePrexInfo(target, NULL);
+   }
+
+/** \brief
+ *     Find the def to an auto or parm before treetop in a extended basic block
+ *
+ *  \return
+ *     The treetop containing the def (the store)
+ */
+static TR::TreeTop*
+defToAutoOrParmInEBB(TR::Compilation* comp, TR::TreeTop* treetop, TR::SymbolReference* symRef, TR::Node** valueNode)
+   {
+   while (treetop)
+      {
+      auto ttNode = treetop->getNode();
+      if (ttNode->getOpCodeValue() == TR::BBStart)
+         {
+         auto block = ttNode->getBlock();
+         if (!block->isExtensionOfPreviousBlock())
+            return NULL;
+         else
+            continue;
+         }
+
+      auto storeNode = ttNode->getStoreNode();
+      if (storeNode &&
+          storeNode->getOpCode().isStoreDirect() &&
+          storeNode->getSymbolReference() == symRef)
+         {
+         auto child = storeNode->getFirstChild();
+         // If the child is also an auto, keep walking the trees to find the child's def
+         if (child->getOpCode().hasSymbolReference() &&
+             child->getSymbolReference()->getSymbol()->isAuto() &&
+             !child->getSymbolReference()->hasKnownObjectIndex())
+            {
+            symRef = child->getSymbolReference();
+            continue;
```

Comment:
Do you need to explicitly advance the treetop to the prev one before this continue is executed ?

Reply:
Right, fixed the bug and modified some trace messages

Reply:
I cannot see the prior version of the commit containing the code at this spot but I wanted to ask if the change from a while loop to a for loop that you made in your last change will work as you intended for the other continue statement(s) in the loop. i.e. changing from a while to a for loop works better for the case that I asked about but I wanted to be clear that it does not change behaviour for the other continue(s) in the same loop.

Reply:
Yes, the for loop will advance the treetop in each iteration, thus I don't need to update the treetop before `continue`. The code will work as intended.